db/src/migration.rs

```diff
@@ -0,0 +1,75 @@
+use crate::{db::VERSION_KEY, internal_error, Result};
+use ckb_logger::info;
+use rocksdb::{
+    ops::{Get, Put},
+    OptimisticTransactionDB,
+};
+
+#[derive(Default)]
+pub struct Migrations {
+    migrations: Vec<Box<dyn Migration>>,
+}
+
+impl Migrations {
+    pub fn add_migration(&mut self, migration: Box<dyn Migration>) {
+        self.migrations.push(migration);
+    }
+
+    pub fn migrate(&self, db: &OptimisticTransactionDB) -> Result<()> {
+        let db_version = db
+            .get(VERSION_KEY)
+            .map_err(|err| {
+                internal_error(format!("failed to get the version of database: {}", err))
+            })?
+            .map(|version_bytes| unsafe { String::from_utf8_unchecked(version_bytes.to_vec()) });
+
+        match db_version {
+            Some(v) => self
+                .migrations
+                .iter()
+                .filter(|m| m.version() > v.as_str())
+                .try_for_each(|m| {
+                    info!("Migrating database version to {}", m.version());
+                    m.migrate(db)
+                }),
+            None => {
+                if let Some(m) = self.migrations.last() {
+                    info!("Init database version {}", m.version());
+                    db.put(VERSION_KEY, m.version()).map_err(|err| {
+                        internal_error(format!("failed to migrate the database: {}", err))
+                    })
+                } else {
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+pub trait Migration {
```

Comment:
There are versions we cannot do migrations, such as from "0.1.0" to "1.0.0". Another issue is that this PR accepts an old database such as "0.2000.0" as a compatible version and happily upgrades to "0.2500.1".

I thinks the trait should have a method `upgradable(&self, current_version: Option<&str>) -> bool`  to check whether it is possible to do a migration. The default implementation can use semver to decide whether it is possible to migrate the database.